document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -263,8 +269,166 @@ public BulkUpdateResult bulkUpdateSubDocs(Map<Key, Map<String, Document>> docume
   }
 
   @Override
-  public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest request) {
-    throw new UnsupportedOperationException();
+  public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest request)
+      throws Exception {
+    Operation operation = request.getOperation();
+    switch (operation) {
+        // todo: All of these implementations do a GET -> in memory update -> UPSERT which is not
+        // optimal. Changes this logic to use a single query using jsob_set and json_insert
+      case ADD:
+        return addImpl(request);
+      case SET:
+        return setImpl(request);
+      case REMOVE:
+        return removeImpl(request);
+      default:
+        throw new UnsupportedOperationException("Unsupported operation: " + operation);
+    }
+  }
+
+  private BulkUpdateResult removeImpl(BulkArrayValueUpdateRequest request) throws IOException {
+    Set<Key> keys = request.getKeys();
+    String[] pathTokens = request.getSubDocPath().split("\\.");
+    Map<Key, Document> upsertMap = new HashMap<>();
+    for (Key key : keys) {
+      Query getQuery = getSelectDocumentQueryForKey(key);
```

Comment:
We can use a single Query to fetch all documents using the `IN` filter?

```
Select document from <> where ID in (<v1>, <v2>,... )
```

Any specific reason for doing iteratively?

Reply:
Missed this, will fix.